File: 21551-naveen-siva-kumar/week7/2420-find-all-good-indices/2420-find-all-good-indices.cpp
class Solution {
public:
    vector<int> goodIndices(vector<int>& nums, int k) {
        
        int N = nums.size();
        vector<int> decreasing(N, 1), increasing(N, 1);
        vector<int> ans;
        
        for(int i=1;i<N;i++){
            if(nums[i]<=nums[i-1]){
                decreasing[i] = decreasing[i-1]+1;
            }
        }
        for(int i = N-2;i>-1;i--){
            if(nums[i]<=nums[i+1]){
                increasing[i] = increasing[i+1]+1;
            }
        }
        
        for(int i=k;i<N-k;i++){
            if(decreasing[i-1]>=k && increasing[i+1]>=k ){
                ans.push_back(i);
            }
        }
        return ans;
    }
};

File: 21553 - Sai Nandan D/Week 7/2420_Find All Good Indices.py
class Solution:
    def goodIndices(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        non_inc, non_dec = [1]*n, [1]*n
        
        for i in range(1, n):
            if not (nums[i] > nums[i-1]): non_inc[i] = non_inc[i-1] + 1
            if not (nums[~i+1] < nums[~i]): non_dec[~i] = non_dec[~i+1] + 1
                
        return [i for i in range(k, n-k) if non_inc[i-1] >= k and non_dec[i+1] >= k]
